utils_nlp/models/transformers/common.py

```diff
@@ -1,28 +1,50 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+
+# This script reuses some code from
+# https://github.com/huggingface/pytorch-transformers/blob/master/examples/run_glue.py
+
+
+from pytorch_transformers.modeling_bert import BERT_PRETRAINED_MODEL_ARCHIVE_MAP
+from pytorch_transformers.modeling_xlnet import XLNET_PRETRAINED_MODEL_ARCHIVE_MAP
+from pytorch_transformers.modeling_roberta import ROBERTA_PRETRAINED_MODEL_ARCHIVE_MAP
+from pytorch_transformers.modeling_distilbert import DISTILBERT_PRETRAINED_MODEL_ARCHIVE_MAP
+from pytorch_transformers import (
+    BertTokenizer,
+    DistilBertTokenizer,
+    RobertaTokenizer,
+    XLNetTokenizer,
+)
+
+TOKENIZER_CLASS = {}
+TOKENIZER_CLASS.update({k: BertTokenizer for k in BERT_PRETRAINED_MODEL_ARCHIVE_MAP})
+TOKENIZER_CLASS.update({k: RobertaTokenizer for k in ROBERTA_PRETRAINED_MODEL_ARCHIVE_MAP})
+TOKENIZER_CLASS.update({k: XLNetTokenizer for k in XLNET_PRETRAINED_MODEL_ARCHIVE_MAP})
+TOKENIZER_CLASS.update({k: DistilBertTokenizer for k in DISTILBERT_PRETRAINED_MODEL_ARCHIVE_MAP})
+
+MAX_SEQ_LEN = 512
 
 
 def fine_tune(
     model,
     model_type,
     train_dataloader,
+    num_epochs,
+    device,
     n_gpu,
     local_rank,
-    device,
-    gradient_accumulation_steps,    
+    gradient_accumulation_steps,
     fp16,
     seed,
 ):
     global_step = 0
     tr_loss = 0.0
     model.zero_grad()
-    train_iterator = trange(
-        int(num_epochs), desc="Epoch", disable=local_rank not in [-1, 0]
-    )
+    train_iterator = trange(int(num_epochs), desc="Epoch", disable=local_rank not in [-1, 0])
     set_seed(seed)
 
     for _ in train_iterator:
-        epoch_iterator = tqdm(
-            train_dataloader, desc="Iteration", disable=local_rank not in [-1, 0]
-        )
+        epoch_iterator = tqdm(train_dataloader, desc="Iteration", disable=local_rank not in [-1, 0])
         for step, batch in enumerate(epoch_iterator):
             model.train()
             batch = tuple(t.to(device) for t in batch)
```

Comment:
lines 51 - 56 will not work for QA. For QA the inputs are 
                inputs = {
                    "input_ids": batch[0],
                    "attention_mask": batch[1],
                    "token_type_ids": batch[2],
                    "start_positions": batch[3],
                    "end_positions": batch[4],
                }
                if self.model_type in ["xlnet"]:
                    inputs.update({"cls_index": batch[5], "p_mask": batch[6]})

How about we create a TransformerBase class and let each class implements its own construct_inputs method? The base class can also have a create_optimizer method that's shared across all classes.

Reply:
Yes, good idea. We only need 2 for now, QA format and the rest. Anything else?

Reply:
token classification doesn't have token_type_ids at the moment, but we can add it. So only 2 for now, QA and the rest.